src/mono/mono/mini/interp/interp-simd.c

```diff
@@ -0,0 +1,566 @@
+}
+
+static void
+interp_v128_i2_create_scalar (gpointer res, gpointer v1)
+{
+	gint16 val = *(gint16*)v1;
+	memset (res, 0, SIZEOF_V128);
+	*(gint16*)res = val;
+}
+static void
+interp_v128_i4_create_scalar (gpointer res, gpointer v1)
+{
+	gint32 val = *(gint32*)v1;
+	memset (res, 0, SIZEOF_V128);
+	*(gint32*)res = val;
+}
+static void
+interp_v128_i8_create_scalar (gpointer res, gpointer v1)
+{
+	gint64 val = *(gint64*)v1;
+	memset (res, 0, SIZEOF_V128);
+	*(gint64*)res = val;
+}
+
+// ExtractMostSignificantBits
+static void
+interp_v128_i1_extract_msb (gpointer res, gpointer v1)
+{
+	guint32 val = 0;
+	gint8 *v1_typed = (gint8*)v1;
+	for (int i = 0; i < SIZEOF_V128 / sizeof (gint8); i++) {
+		if (*v1_typed & (1 << 7))
+			val |= 1 << i;
+		v1_typed++;
+	}
+	*(guint32*)res = val;
+}
+
+static void
+interp_v128_i2_extract_msb (gpointer res, gpointer v1)
+{
+	guint32 val = 0;
+	gint16 *v1_typed = (gint16*)v1;
+	for (int i = 0; i < SIZEOF_V128 / sizeof (gint16); i++) {
+		if (*v1_typed & (1 << 15))
+			val |= 1 << i;
+		v1_typed++;
+	}
+	*(guint32*)res = val;
+}
+
+static void
+interp_v128_i4_extract_msb (gpointer res, gpointer v1)
+{
+	guint32 val = 0;
+	gint32 *v1_typed = (gint32*)v1;
+	for (int i = 0; i < SIZEOF_V128 / sizeof (gint32); i++) {
+		if (*v1_typed & (1 << 31))
+			val |= 1 << i;
+		v1_typed++;
+	}
+	*(guint32*)res = val;
+}
+
+static void
+interp_v128_i8_extract_msb (gpointer res, gpointer v1)
+{
+	guint32 val = 0;
+	gint64 *v1_typed = (gint64*)v1;
+	for (int i = 0; i < SIZEOF_V128 / sizeof (gint64); i++) {
+		if (*v1_typed & ((gint64)1 << 63))
+			val |= 1 << i;
+		v1_typed++;
+	}
+	*(guint32*)res = val;
+}
+
+// ConditionalSelect
+static void
+interp_v128_conditional_select (gpointer res, gpointer v1, gpointer v2, gpointer v3)
+{
+	v128_i8 cond = *(v128_i8*)v1;
+	*(v128_i8*)res = (*(v128_i8*)v2 & cond) | (*(v128_i8*)v3 & ~cond);
+}
+
+// Create
+static void
+interp_v128_i1_create (gpointer res, gpointer v1)
+{
+	gint8 val = *(gint8*)v1;
+	v128_i1 v = { val, val, val, val,
+			val, val, val, val,
+			val, val, val, val,
+			val, val, val, val };
+	*(v128_i1*)res = v;
+}
+
+static void
+interp_v128_i2_create (gpointer res, gpointer v1)
+{
+	gint16 val = *(gint16*)v1;
+	v128_i2 v = { val, val, val, val,
+			val, val, val, val };
+	*(v128_i2*)res = v;
+}
+
+static void
+interp_v128_i4_create (gpointer res, gpointer v1)
+{
+	gint32 val = *(gint32*)v1;
+	v128_i4 v = { val, val, val, val };
+	*(v128_i4*)res = v;
+}
+
+static void
+interp_v128_i8_create (gpointer res, gpointer v1)
+{
+	gint64 val = *(gint64*)v1;
+	v128_i8 v = { val, val };
+	*(v128_i8*)res = v;
+}
+
+// AndNot
+static void
+interp_v128_and_not (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_i4*)res = *(v128_i4*)v1 & ~(*(v128_i4*)v2);
+}
+
+// LessThanOrEqual
+static void
+interp_v128_u2_less_than_equal (gpointer res, gpointer v1, gpointer v2)
+{
+	*(v128_u2*)res = *(v128_u2*)v1 <= *(v128_u2*)v2;
+}
+
+// Shuffle
+
+#define V128_SHUFFLE(eltype, itype) do {					\
```

Comment:
Curious why you have to do this by hand and not with `__builtin_shuffle` ?

Reply:
I simply didn't want to make use of any additional builtins for now, to keep it as simple as possible and reduce potential compatibility issues. Will start adding various builtins in future PR's, probably even arch specific ones. I think `__builtin_shufflevector` has better support and is very powerful, so might be a better candidate.